src/StreamJsonRpc/MessageFormatterHelper.cs

```diff
@@ -0,0 +1,126 @@
+﻿
+namespace StreamJsonRpc
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using Microsoft;
+    using Microsoft.VisualStudio.Threading;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// Class containing useful method to help on the implementation of message formatters.
+    /// </summary>
+    public class MessageFormatterHelper
+    {
+        /// <summary>
+        /// Special method name for progress notification.
+        /// </summary>
+        protected const string ProgressRequestSpecialMethod = "$/progress";
+
+        /// <summary>
+        /// Object used to lock the acces to <see cref="requestProgressMap"/> and <see cref="progressMap"/>.
+        /// </summary>
+        protected readonly object progressLock = new object();
+
+        /// <summary>
+        /// Dictionary used to map the request id to their progress id token so that the progress objects are cleaned after getting the final response.
+        /// </summary>
+        protected readonly Dictionary<long, long> requestProgressMap = new Dictionary<long, long>();
+
+        /// <summary>
+        /// Dictionary used to map progress id token to its corresponding ProgressParamInformation instance containing the progress object and the necessary fields to report the results.
```

Comment:
>ProgressParamInformation [](start = 74, length = 24)

Use `<see cref="ProgressParamInformation" />` here to help ensure the comment stays current if these types are renamed. It also helps with xml doc comments when they are converted to HTML by creating hyperlinks.